### ISSUE TITLE
Option to disk read in do() loop

### ISSUE THREAD

**Issue by rajnisheu** [cubiclesoft/service-manager-src#2]
Landed up here from SO. Thanks for posting the solution as I guess it can save tons of programming time!

Am planning to use this for PHP. From the sample test_service.php, The main loop `do()` is where I would  (probably?) write some socket stuff.

However, I am a bit anxious that the routine does a disk read every `sleep(1);`. I am wondering if there is a better way to do it.

One option could be to use global variables and turn them on/off from the start/stop/pause routines and test them inside the `do()`.

What do you think? Maybe I missed something which needs those files?

**Comment by cubiclesoft** [comment]
The Service Manager binaries create the files that the code in the do loop is checking for.

In the test service source code, the `sleep(1)` is what you replace with your socket code.  The service checking code only runs once every 3 seconds.  Checking for the existence of a file is extremely negligible as long as it isn't a file on a NFS, Samba, or other remote file system share.  The OS' built in file system cache knows whether a file exists or not on the system's own devices without ever querying those devices beyond the first check (i.e. each check does invoke a kernel mode switch but the operation itself happens entirely in RAM).  As a result, checking for a file's existence (and obtaining basic stats) is a pretty cheap operation.  Actually opening the file to read it is much more costly.  You don't really have anything to worry about here.

I use Service Manager in various projects.  I haven't encountered any issues in production and it works like a champ.  Check out the source code to [Cloud Storage Server](https://github.com/cubiclesoft/cloud-storage-server) for an example integration that uses sockets.

**Comment by rajnisheu** [comment]
Thanks for the example mate. Unfortunately, I'm finding it a bit hard to follow the example with a bunch of pointers to other classes and files. I tried this simple insertion but hit the first road block with a surprising php error `Call to undefined function socket_create()`. Looks like I'm missing something basic.

This is how I've started off the stuff:

    ...
	else if ($argv[1] == "start")
	{
		// Start the service.
		$result = $sm->Start($servicename, true);
		if (!$result["success"])  DisplayError("Unable to start the '" . $servicename . "' service.", $result);
		if ($result["success"]) {
					// socket it
			$host = "127.0.0.1";
			$port = 8000;
				// don't timeout?
			//set_time_limit(0);
				// create socket
			$socket = socket_create(AF_INET, SOCK_STREAM, 0);
				// bind socket to port
			$jlt = socket_bind($socket, $host, $port);
				// start listening for connections
			$jlt = socket_listen($socket, 3);
			echo "Server up. Listening\n";
		}
	}
	else if ($argv[1] == "stop")
	{
		// Stop the service.
		$result = $sm->Stop($servicename, true);
		if (!$result["success"])  DisplayError("Unable to stop the '" . $servicename . "' service.", $result);
		if ($result["success"]) {
				// close the socket
			socket_close($socket);
		}
    }

Any idea what could be going on?

**Comment by cubiclesoft** [comment]
The `socket_` PHP functions aren't always compiled into PHP and are not as stable as the `stream_` functions.

You should probably write your TCP/IP server long before you attempt to integrate it with Service Manager.  At this point, integrating with system services should be the last thing on your mind.  Know that it is possible and come back to it when you are ready.

You shouldn't need to change the 'install', 'start', 'stop', etc. Service Manager SDK integration code unless you want multiple services under different names using the same code base.  The only change to the code should be setting the `$servicename` variable and replacing `sleep(1)` with your final application logic.

Issue trackers are for submitting bug reports about bugs in the software and helping the author out by identifying how to reliably trigger the bug.  There are a ton of tutorials on the topic of writing socket servers, so you should probably find one of those.

**Comment by rajnisheu** [comment]
Thanks for your comment.
For whatever reason, I had to reboot the machine after which php stopped throwing the error. (Haven't seen that one before.)
At this point I am still trying to establish features of service-manager and the question is less about php, sockets or streams, but more about general usage of service-manager and establishing how to use it correctly.
You are right this _is_ for bug reporting. Maybe you would like to point me to a location where I could discuss usage questions? You may like to tag this thread appropriately

**Comment by cubiclesoft** [comment]
General discussion forums are here:  http://barebonescms.com/forums/forumdisplay.php?fid=2

I think the simplest way to view Service Manager is that it is the ultimate system service startup helper application.  It's cross-platform so it'll run just about everywhere and acts as middleware between the OS and your application.  It manages the nuances of system services across all major desktop and server OSes so that you don't have to.  The only tricky part is installing your application, which is where a SDK comes in handy BUT you can also install an application by correctly calling the Service Manager binary yourself, albeit the `install` command line gets a tad unwieldy.
